### PR TITLE
[CI TEST] Exp/python310+withlogic+numpy121 v2

### PR DESCRIPTION
**Pull request by esc**
Part of testing #7562 

Integration of the following branches for testing purposes:

#7557
#7498
#7560
#7483

### CODE REVIEW

**Comment by esc** [comment]
Testing this locally on a Mac with Cataline 10.15.07 Intel processor Python 3.10 via Anaconda and Numpy 1.21.4 via conda-forge. Yields:

```
======================================================================
ERROR: test_min_max_iterable_input (numba.tests.test_builtins.TestBuiltins)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/vhaenel/git/numba/numba/core/errors.py", line 817, in new_error_context
    yield
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 280, in lower_block
    self.lower_inst(inst)
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 425, in lower_inst
    val = self.lower_assign(ty, inst)
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 642, in lower_assign
    res = self.lower_yield(ty, value)
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 652, in lower_yield
    y.lower_yield_suspend()
  File "/Users/vhaenel/git/numba/numba/core/generators.py", line 323, in lower_yield_suspend
    val = self.lower.loadvar(name)
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 1388, in loadvar
    ptr = self.getvar(name)
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 1380, in getvar
    return self.varmap[name]
KeyError: '$36pred'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/vhaenel/git/numba/numba/tests/test_builtins.py", line 1061, in test_min_max_iterable_input
    self._check_min_max(fn)
  File "/Users/vhaenel/git/numba/numba/tests/test_builtins.py", line 1037, in _check_min_max
    expected = pyfunc()
  File "/Users/vhaenel/git/numba/numba/tests/test_builtins.py", line 1056, in <lambda>
    yield lambda: op(frange(1.1, 3.3, 0.1))
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 487, in _compile_for_args
    raise e
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 420, in _compile_for_args
    return_val = self.compile(tuple(argtypes))
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 965, in compile
    cres = self._compiler.compile(args, return_type)
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 125, in compile
    status, retval = self._compile_cached(args, return_type)
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 139, in _compile_cached
    retval = self._compile_core(args, return_type)
  File "/Users/vhaenel/git/numba/numba/core/dispatcher.py", line 152, in _compile_core
    cres = compiler.compile_extra(self.targetdescr.typing_context,
  File "/Users/vhaenel/git/numba/numba/core/compiler.py", line 690, in compile_extra
    return pipeline.compile_extra(func)
  File "/Users/vhaenel/git/numba/numba/core/compiler.py", line 428, in compile_extra
    return self._compile_bytecode()
  File "/Users/vhaenel/git/numba/numba/core/compiler.py", line 496, in _compile_bytecode
    return self._compile_core()
  File "/Users/vhaenel/git/numba/numba/core/compiler.py", line 475, in _compile_core
    raise e
  File "/Users/vhaenel/git/numba/numba/core/compiler.py", line 462, in _compile_core
    pm.run(self.state)
  File "/Users/vhaenel/git/numba/numba/core/compiler_machinery.py", line 346, in run
    raise patched_exception
  File "/Users/vhaenel/git/numba/numba/core/compiler_machinery.py", line 334, in run
    self._runPass(idx, pass_inst, state)
  File "/Users/vhaenel/git/numba/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/Users/vhaenel/git/numba/numba/core/compiler_machinery.py", line 289, in _runPass
    mutated |= check(pss.run_pass, internal_state)
  File "/Users/vhaenel/git/numba/numba/core/compiler_machinery.py", line 262, in check
    mangled = func(compiler_state)
  File "/Users/vhaenel/git/numba/numba/core/typed_passes.py", line 396, in run_pass
    lower.lower()
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 189, in lower
    self.genlower.lower_next_func(self)
  File "/Users/vhaenel/git/numba/numba/core/generators.py", line 152, in lower_next_func
    entry_block_tail = lower.lower_function_body()
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 266, in lower_function_body
    self.lower_block(block)
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 278, in lower_block
    with new_error_context('lowering "{inst}" at {loc}', inst=inst,
  File "/Users/vhaenel/miniconda3/envs/numba_3.10/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/vhaenel/git/numba/numba/core/errors.py", line 831, in new_error_context
    raise newerr.with_traceback(tb)
  File "/Users/vhaenel/git/numba/numba/core/errors.py", line 817, in new_error_context
    yield
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 280, in lower_block
    self.lower_inst(inst)
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 425, in lower_inst
    val = self.lower_assign(ty, inst)
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 642, in lower_assign
    res = self.lower_yield(ty, value)
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 652, in lower_yield
    y.lower_yield_suspend()
  File "/Users/vhaenel/git/numba/numba/core/generators.py", line 323, in lower_yield_suspend
    val = self.lower.loadvar(name)
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 1388, in loadvar
    ptr = self.getvar(name)
  File "/Users/vhaenel/git/numba/numba/core/lowering.py", line 1380, in getvar
    return self.varmap[name]
numba.core.errors.LoweringError: Failed in nopython mode pipeline (step: native lowering)
'$36pred'

File "numba/tests/test_builtins.py", line 1047:
        def frange(start, stop, step):
            <source elided>
            while i < stop:
                yield i
                ^

During: lowering "$18yield_value.1 = yield i.2" at /Users/vhaenel/git/numba/numba/tests/test_builtins.py (1047)

======================================================================
FAIL: test_median_npdatetime (numba.tests.test_array_reductions.TestArrayReductions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/vhaenel/git/numba/numba/tests/test_array_reductions.py", line 729, in test_median_npdatetime
    self.check_nptimedelta(array_median_global)
  File "/Users/vhaenel/git/numba/numba/tests/test_array_reductions.py", line 710, in check_nptimedelta
    self._do_check_nptimedelta(pyfunc, arr)
  File "/Users/vhaenel/git/numba/numba/tests/test_array_reductions.py", line 690, in _do_check_nptimedelta
    self.assertPreciseEqual(cfunc(arr), pyfunc(arr))
  File "/Users/vhaenel/git/numba/numba/tests/support.py", line 366, in assertPreciseEqual
    self.fail("when comparing %s and %s: %s" % (first, second, failure_msg))
AssertionError: when comparing 3 seconds and 4 seconds: numpy.timedelta64(3,'s') != numpy.timedelta64(4,'s')

======================================================================
FAIL: test_DILocation_versioned_variables (numba.tests.test_debuginfo.TestDebugInfoEmission)
Tests that DILocation information for versions of variables matches
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/vhaenel/git/numba/numba/tests/test_debuginfo.py", line 382, in test_DILocation_versioned_variables
    self.assertEqual(len(associated_lines), 2) # 2 SSA versioned names 'c'
AssertionError: 1 != 2

======================================================================
FAIL: test_check_objmode_cache_ndarray (numba.tests.test_extending.TestCachingOverloadObjmode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/vhaenel/git/numba/numba/tests/test_extending.py", line 1860, in test_check_objmode_cache_ndarray
    self.assertEqual(res['exitcode'], 0)
AssertionError: 1 != 0

----------------------------------------------------------------------
Ran 9672 tests in 2613.218s

FAILED (failures=3, errors=1, skipped=742, expected failures=13)
python -m numba.runtests -m 12 -v numba.tests  13996.43s user 1111.90s system 575% cpu 43:44.56 total
```

**Comment by stuartarchibald** [comment]
@esc https://github.com/stuartarchibald/numba/commit/7ecc6520cac3143ad4e37c8ffd70adf835a17149 fixes `test_DILocation_versioned_variables`

**Comment by esc** [comment]
> @esc [stuartarchibald@7ecc652](https://github.com/stuartarchibald/numba/commit/7ecc6520cac3143ad4e37c8ffd70adf835a17149) fixes `test_DILocation_versioned_variables`

Thank you, I have cherry-picked this and moved it onto: e44e0dbea07eee5b0b718e473d9247d6b7455914

**Comment by esc** [comment]
branch has been re-built with latest sub-branches, testing in progress...

**Comment by stuartarchibald** [comment]
This:
```diff
diff --git a/numba/core/lowering.py b/numba/core/lowering.py
index 92cc76d..0f457ad 100644
--- a/numba/core/lowering.py
+++ b/numba/core/lowering.py
@@ -49,7 +49,6 @@ class BaseLower(object):
         self.pending_phis = {}
         self.varmap = {}
         self.firstblk = min(self.blocks.keys())
-        self.loc = -1
 
         # Specializes the target context as seen inside the Lowerer
         # This adds:
@@ -57,13 +56,17 @@ class BaseLower(object):
         self.context = context.subtarget(environment=self.env,
                                          fndesc=self.fndesc)
 
+        # initial location is set to def line
+        defn_loc = self._compute_def_location()
+        self.loc = defn_loc
+
         # Debuginfo
         dibuildercls = (self.context.DIBuilder
                         if self.context.enable_debuginfo
                         else debuginfo.DummyDIBuilder)
 
         # debuginfo def location
-        self.defn_loc = self._compute_def_location()
+        self.defn_loc = defn_loc
 
         self.debuginfo = dibuildercls(module=self.module,
                                       filepath=func_ir.loc.filename,
@@ -377,7 +380,7 @@ class Lower(BaseLower):
         super(Lower, self).pre_block(block)
         self._cur_ir_block = block
 
-        if block == self.firstblk:
+        if block == self.blocks[self.firstblk]:
             # create slots for all the vars, irrespective of whether they are
             # initialized, SSA will pick this up and warn users about using
             # uninitialized variables. Slots are added as alloca in the first
```
will "fix" the `numba.tests.test_builtins.TestBuiltins.test_min_max_iterable_input` test. It basically fixes a bug that was preventing `alloca` slots being written for all variables. I'm not convinced of this yet, I think there's potentially something else going on and I'm also reasonably sure that this will break DWARF.

**Comment by sklam** [comment]
@stuartarchibald, the problem is in this block:

Screen Shot 2021-11-16 at 2 33 28 PM
 
at yield (2nd instruction in the block). The generator lowering expects `$36pred`, which is defined later in the block, to have an alloca slot already. Since that variable is used as the condition, and the block loops back to itself, the expectation on the alloca slot makes sense. So, I believe your patch makes sense.

**Comment by stuartarchibald** [comment]
@sklam thanks for looking closer at the IR, I see what's going on now. I think the fix could be more local (and won't break DWARF as a result). This patch: https://github.com/stuartarchibald/numba/commit/ac4bba3370885f1a2bb531e31ef15d14c889e817 produces a similar effect of ensuring that the live vars in the generator state are `alloca`'d ahead of use but is far less invasive, seems to keep DWARF working too.

**Comment by sklam** [comment]
@stuartarchibald, https://github.com/stuartarchibald/numba/commit/ac4bba3370885f1a2bb531e31ef15d14c889e817 makes sense. 

Btw, I am starting to think we can avoid putting `del` of block-local variable into the successors by forcing predicates to be of type `i1`.

**Comment by stuartarchibald** [comment]
@sklam thanks for confirming. @esc I think https://github.com/stuartarchibald/numba/commit/088b3658706cc2f7b33c5b9a9423e71f4792162a is also needed to fix the issue with objmode and generators.

**Comment by esc** [comment]
will rebuild this branch now

**Comment by esc** [comment]
Octomerge :octocat: worked.

**Comment by stuartarchibald** [comment]
Buildfarm ID: `numba_smoketest_cpu_yaml_62`.

**Comment by esc** [comment]
> Buildfarm ID: `numba_smoketest_cpu_yaml_62`.

After adding new tasks for Python 3.10 -- I discovered they all failed due to lack of availability of packages. For example Numpy, but perhaps also others. I am trying to rebuild this now with `conda-forge` dependencies.

**Comment by esc** [comment]
rebuilt with #7557 and #7560 only since Numpy 1.21 support is now in master with the merge of #7483

**Comment by esc** [comment]
rebuilt again

**Comment by esc** [comment]
`Build numba_smoketest_cpu_yaml_66 has started`

**Comment by esc** [comment]
Screen Shot 2021-11-24 at 19 49 01

This was fine, given the existing holes in the matrix caused by lack of Numpy packages (any Numpy packages) for 3.10 via the `defaults`/`main` channel.

**Comment by esc** [comment]
Rebuilt, but w/o Numpy 1.21 (branch #7483) since this is in `master` now.

**Comment by esc** [comment]
I added a hack for our public CI setup to allow testing of 3.10. Here is the patch for safe keeping, since this branch is subject to rebase:

```
diff --git a/azure-pipelines.yml b/azure-pipelines.yml
index 6b1530c673..d2f9f80571 100644
--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -3,7 +3,7 @@ trigger:

 variables:
   # Change the following along with adding new TEST_START_INDEX.
-  TEST_COUNT: 19
+  TEST_COUNT: 20

 jobs:
 # Mac and Linux use the same template with different matrixes
@@ -116,6 +116,11 @@ jobs:
         NUMPY: '1.21'
         CONDA_ENV: azure_ci
         TEST_START_INDEX: 16
+      py310_np121:
+        PYTHON: '3.10'
+        NUMPY: '1.21'
+        CONDA_ENV: azure_ci
+        TEST_START_INDEX: 17

 - template: buildscripts/azure/azure-windows.yml
   parameters:
diff --git a/buildscripts/azure/azure-windows.yml b/buildscripts/azure/azure-windows.yml
index eff6a6655d..eb929088a6 100644
--- a/buildscripts/azure/azure-windows.yml
+++ b/buildscripts/azure/azure-windows.yml
@@ -12,12 +12,12 @@ jobs:
         PYTHON: '3.9'
         NUMPY: '1.21'
         CONDA_ENV: 'testenv'
-        TEST_START_INDEX: 17
+        TEST_START_INDEX: 18
       py37_np118:
         PYTHON: '3.7'
         NUMPY: '1.18'
         CONDA_ENV: 'testenv'
-        TEST_START_INDEX: 18
+        TEST_START_INDEX: 19

   steps:
     - task: CondaEnvironment@1
diff --git a/buildscripts/incremental/setup_conda_environment.sh b/buildscripts/incremental/setup_conda_environment.sh
index b256a9a9ff..4236b1a14a 100755
--- a/buildscripts/incremental/setup_conda_environment.sh
+++ b/buildscripts/incremental/setup_conda_environment.sh
@@ -40,7 +40,11 @@ conda list
 # NOTE: pyyaml is used to ensure that the Azure CI config is valid
 # NOTE: 32 bit linux... do not install NumPy, there's no conda package for >1.15
 # so it has to come from pip later
-if [[ "$CONDA_SUBDIR" == "linux-32" || "$BITS32" == "yes" ]]; then
+# If it's Python 3.10, we get everything except for the interpreter and
+# compilers via pip from PyPi
+if [[ "$PYTHON" == "3.10" ]] ; then
+    conda create -n $CONDA_ENV -q -y ${EXTRA_CHANNELS} python=$PYTHON
+elif [[ "$CONDA_SUBDIR" == "linux-32" || "$BITS32" == "yes" ]] ; then
     conda create -n $CONDA_ENV -q -y ${EXTRA_CHANNELS} python=$PYTHON pip gitpython pyyaml
 else
     conda create -n $CONDA_ENV -q -y ${EXTRA_CHANNELS} python=$PYTHON numpy=$NUMPY pip gitpython pyyaml
@@ -52,7 +56,7 @@ source activate $CONDA_ENV
 set -v

 # Install optional packages into activated env
-if [ "${VANILLA_INSTALL}" != "yes" ]; then
+if [ "$PYTHON" != "3.10" || "${VANILLA_INSTALL}" != "yes" ]; then
     # Scipy, CFFI, jinja2, IPython and pygments are optional dependencies, but exercised in the test suite
     # pexpect is used to run the
     $CONDA_INSTALL ${EXTRA_CHANNELS} cffi jinja2 ipython pygments pexpect
@@ -76,8 +80,13 @@ elif  [[ $(uname) == Darwin ]]; then
     $CONDA_INSTALL llvm-openmp intel-openmp
 fi

+# `pip install` the dependencies on Python 3.10
+if [[ "$PYTHON" == "3.10" ]] ; then
+    $PIP_INSTALL -U pip
+    pip --version
+    $PIP_INSTALL gitpython pyyaml cffi jinja2 ipython pygments pexpect scipy numpy
 # If on 32bit linux, now pip install NumPy (no conda package), SciPy is broken?!
-if [[ "$CONDA_SUBDIR" == "linux-32" || "$BITS32" == "yes" ]] ; then
+elif [[ "$CONDA_SUBDIR" == "linux-32" || "$BITS32" == "yes" ]] ; then
     $PIP_INSTALL numpy==$NUMPY
 fi
```

**Comment by esc** [comment]
Screen Shot 2021-12-03 at 17 25 49
 
hmmm ....

File: buildscripts/incremental/setup_conda_environment.sh [hunk 1]
```diff
@@ -52,7 +56,7 @@ source activate $CONDA_ENV
 set -v
 
 # Install optional packages into activated env
-if [ "${VANILLA_INSTALL}" != "yes" ]; then
+if [ "$PYTHON" != "3.10" || "${VANILLA_INSTALL}" != "yes" ]; then
```
Comment (stuartarchibald) [inline buildscripts/incremental/setup_conda_environment.sh:59]: Probably needs to be:
```suggestion
if [ "$PYTHON" != "3.10" ] || [ "${VANILLA_INSTALL}" != "yes" ]; then
```
or
```suggestion
if [[ "$PYTHON" != "3.10" || "${VANILLA_INSTALL}" != "yes" ]]; then
```

Comment (esc) [reply buildscripts/incremental/setup_conda_environment.sh]: The solution was:

```
if [[ "$PYTHON" != "3.10" && "${VANILLA_INSTALL}" != "yes" ]]; then
```

**Comment by esc** [comment]
here is the final variant of the patch:

```
commit 630ccf933f5176f591fc2a284371f35e1cb238c7 (HEAD -> exp/python310+withlogic+numpy121_v2, esc/exp/python310+withlogic+numpy121_v2)
Author: esc <esc@users.noreply.github.com>
Date:   Fri Dec 3 11:12:21 2021 +0100

    start testing Python 3.10 on public CI

    As title

diff --git a/azure-pipelines.yml b/azure-pipelines.yml
index 6b1530c673..d2f9f80571 100644
--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -3,7 +3,7 @@ trigger:

 variables:
   # Change the following along with adding new TEST_START_INDEX.
-  TEST_COUNT: 19
+  TEST_COUNT: 20

 jobs:
 # Mac and Linux use the same template with different matrixes
@@ -116,6 +116,11 @@ jobs:
         NUMPY: '1.21'
         CONDA_ENV: azure_ci
         TEST_START_INDEX: 16
+      py310_np121:
+        PYTHON: '3.10'
+        NUMPY: '1.21'
+        CONDA_ENV: azure_ci
+        TEST_START_INDEX: 17

 - template: buildscripts/azure/azure-windows.yml
   parameters:
diff --git a/buildscripts/azure/azure-windows.yml b/buildscripts/azure/azure-windows.yml
index eff6a6655d..eb929088a6 100644
--- a/buildscripts/azure/azure-windows.yml
+++ b/buildscripts/azure/azure-windows.yml
@@ -12,12 +12,12 @@ jobs:
         PYTHON: '3.9'
         NUMPY: '1.21'
         CONDA_ENV: 'testenv'
-        TEST_START_INDEX: 17
+        TEST_START_INDEX: 18
       py37_np118:
         PYTHON: '3.7'
         NUMPY: '1.18'
         CONDA_ENV: 'testenv'
-        TEST_START_INDEX: 18
+        TEST_START_INDEX: 19

   steps:
     - task: CondaEnvironment@1
diff --git a/buildscripts/incremental/setup_conda_environment.sh b/buildscripts/incremental/setup_conda_environment.sh
index b256a9a9ff..038e8a578c 100755
--- a/buildscripts/incremental/setup_conda_environment.sh
+++ b/buildscripts/incremental/setup_conda_environment.sh
@@ -40,7 +40,11 @@ conda list
 # NOTE: pyyaml is used to ensure that the Azure CI config is valid
 # NOTE: 32 bit linux... do not install NumPy, there's no conda package for >1.15
 # so it has to come from pip later
-if [[ "$CONDA_SUBDIR" == "linux-32" || "$BITS32" == "yes" ]]; then
+# If it's Python 3.10, we get everything except for the interpreter and
+# compilers via pip from PyPi
+if [[ "$PYTHON" == "3.10" ]] ; then
+    conda create -n $CONDA_ENV -q -y ${EXTRA_CHANNELS} python=$PYTHON
+elif [[ "$CONDA_SUBDIR" == "linux-32" || "$BITS32" == "yes" ]] ; then
     conda create -n $CONDA_ENV -q -y ${EXTRA_CHANNELS} python=$PYTHON pip gitpython pyyaml
 else
     conda create -n $CONDA_ENV -q -y ${EXTRA_CHANNELS} python=$PYTHON numpy=$NUMPY pip gitpython pyyaml
@@ -52,7 +56,9 @@ source activate $CONDA_ENV
 set -v

 # Install optional packages into activated env
-if [ "${VANILLA_INSTALL}" != "yes" ]; then
+echo "PYTHON=$PYTHON"
+echo "VANILLA_INSTALL=$VANILLA_INSTALL"
+if [[ "$PYTHON" != "3.10" && "$VANILLA_INSTALL" != "yes" ]]; then
     # Scipy, CFFI, jinja2, IPython and pygments are optional dependencies, but exercised in the test suite
     # pexpect is used to run the
     $CONDA_INSTALL ${EXTRA_CHANNELS} cffi jinja2 ipython pygments pexpect
@@ -76,8 +82,13 @@ elif  [[ $(uname) == Darwin ]]; then
     $CONDA_INSTALL llvm-openmp intel-openmp
 fi

+# `pip install` the dependencies on Python 3.10
+if [[ "$PYTHON" == "3.10" ]] ; then
+    $PIP_INSTALL -U pip
+    pip --version
+    $PIP_INSTALL gitpython pyyaml cffi jinja2 ipython pygments pexpect scipy numpy
 # If on 32bit linux, now pip install NumPy (no conda package), SciPy is broken?!
-if [[ "$CONDA_SUBDIR" == "linux-32" || "$BITS32" == "yes" ]] ; then
+elif [[ "$CONDA_SUBDIR" == "linux-32" || "$BITS32" == "yes" ]] ; then
     $PIP_INSTALL numpy==$NUMPY
 fi
```

**Comment by esc** [comment]
#7560 was merged and #7557 was rebased onto master. I have cherry-picked 🍒 the final patch to enable the testing on Python 3.10 using the `pip` installed dependencies and ported it onto the rebased variant of #7557. This means, #7557 remains the only PR related to Python 3.10 support and hence this integration PR can be closed and abandoned.

**Pull request closed without merging**